File: python/lombokalgoritma/graph/kruskal.py
# LombokAlgoritma — Kruskal minimum spanning forest (SPEC §9.8)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from ..datastructure import DisjointSet
from .types import Edge, Graph, validate_graph


def kruskal(g: Graph) -> list[Edge]:
    """Minimum spanning forest (edges read undirected), stable by weight, in the order taken."""
    validate_graph(g)
    ds = DisjointSet(int(g.nodes))
    out: list[Edge] = []
    for u, v, w in sorted(g.edges, key=lambda e: e[2]):
        if ds.union(u, v):
            out.append(Edge(u, v, w))
    return out

File: rust/lombokalgoritma/src/graph/kruskal.rs
// LombokAlgoritma — Kruskal minimum spanning forest (SPEC §9.8)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::types::{validate, Edge, Graph};
use crate::datastructure::DisjointSet;
use crate::sort::mergesort_by;
use crate::Result;
use alloc::vec::Vec;
use core::cmp::Ordering;

/// Minimum spanning forest of the graph read as undirected: edges in a **stable** ascending
/// order of weight (ties keep edge-list order), taken when they join two components. Edges are
/// returned in the order taken, with their original orientation. O(E log E).
///
/// # Errors
/// [`crate::Error::OutOfRange`] for an edge endpoint outside `[0, n)`.
pub fn kruskal(g: &Graph) -> Result<Vec<Edge>> {
    validate(g)?;
    let mut order: Vec<Edge> = g.edges.clone();
    mergesort_by(&mut order, |a, b| {
        a.weight.partial_cmp(&b.weight).unwrap_or(Ordering::Equal)
    });
    let mut ds = DisjointSet::new(g.nodes);
    let mut out = Vec::new();
    for e in order {
        if ds.union(e.from, e.to)? {
            out.push(e);
        }
    }
    Ok(out)
}
